### DISCUSSION TITLE
New install, one camera, some performance issues

### DISCUSSION THREAD

**Discussion by Drizzt321** [blakeblackshear/frigate#4621] [Ask A Question]
So this is a new install into my HA instance, which is running as a VM (HassOS image) on my 1700X FreeBSD host NAS (running via bhyve virtualization). This is via the Add-On + integration, not a separate VM running Frigate + integration. VM has 4 vCPUs in total, 4GB RAM total. 

I have 1 camera, a [SV3C/B08POE-5MPL](https://www.amazon.com/SV3C-Megapixels-Surveillance-Detection-Waterproof/dp/B07PFNQ98P) NVIF camera, 

The biggest issue I'm encountering is that when I go to the Frigate Web UI, and click on the camera to view the live image, I get nothing for several seconds, and then I get what seems to be a static image. And then, based on the camera timestamp in the corner, only occasional frame updates several seconds apart usually. On the host system, I'm really not using hardly any CPU at all, so I'm not sure what's going on.

Secondarily, I'm sure my detection is running rather slow, the inference speed is coming in at 78.97, although the camera process FPS is coming in at the same as the camera FPS, 5.1. But the detection FPS is 0. I don't have Coral TPU yet, I'm trying to find a PCIe one, but seems nearly impossible right now :(  My CPU detector is set to use up to 2 threads. My overall Debug Config is below, if it helps any.

So obviously hardware upgrades would be the Coral AI, I might just have to buy a USB off of that well known auction site and get a USB PCIe card (bhyve only supports PCIe passthrough, not individual USB device passthrough). 

Otherwise, since I have now HW video decoding, a 5700G might work well, passing through the iGPU to the VM in order to access the HW decoding blocks. Also would be a big CPU performance jump as well.

Thoughts? Any config tweaks I can make?

```
{
  "birdseye": {
    "enabled": true,
    "height": 720,
    "mode": "objects",
    "quality": 8,
    "width": 1280
  },
  "cameras": {
    "alley_1": {
      "best_image_timeout": 60,
      "birdseye": {
        "enabled": true,
        "mode": "objects"
      },
      "detect": {
        "enabled": true,
        "fps": 5,
        "height": 1080,
        "max_disappeared": 25,
        "stationary": {
          "interval": 0,
          "max_frames": {
            "default": null,
            "objects": {}
          },
          "threshold": 50
        },
        "width": 1920
      },
      "ffmpeg": {
        "global_args": [
          "-hide_banner",
          "-loglevel",
          "warning"
        ],
        "hwaccel_args": [],
        "input_args": [
          "-avoid_negative_ts",
          "make_zero",
          "-fflags",
          "+genpts+discardcorrupt",
          "-rtsp_transport",
          "tcp",
          "-timeout",
          "5000000",
          "-use_wallclock_as_timestamps",
          "1"
        ],
        "inputs": [
          {
            "global_args": [],
            "hwaccel_args": [],
            "input_args": [],
            "path": "rtsp://admin:<PW>@10.30.255.244/stream0",
            "roles": [
              "record",
              "rtmp",
              "detect"
            ]
          }
        ],
        "output_args": {
          "detect": [
            "-f",
            "rawvideo",
            "-pix_fmt",
            "yuv420p"
          ],
          "record": [
            "-f",
            "segment",
            "-segment_time",
            "10",
            "-segment_format",
            "mp4",
            "-reset_timestamps",
            "1",
            "-strftime",
            "1",
            "-c",
            "copy",
            "-an"
          ],
          "rtmp": [
            "-c",
            "copy",
            "-f",
            "flv"
          ]
        }
      },
      "ffmpeg_cmds": [
        {
          "cmd": "ffmpeg -hide_banner -loglevel warning -avoid_negative_ts make_zero -fflags +genpts+discardcorrupt -rtsp_transport tcp -timeout 5000000 -use_wallclock_as_timestamps 1 -i rtsp://admin:<PW>@10.30.255.244/stream0 -f segment -segment_time 10 -segment_format mp4 -reset_timestamps 1 -strftime 1 -c copy -an /tmp/cache/alley_1-%Y%m%d%H%M%S.mp4 -c copy -f flv rtmp://127.0.0.1/live/alley_1 -r 5 -s 1920x1080 -f rawvideo -pix_fmt yuv420p pipe:",
          "roles": [
            "record",
            "rtmp",
            "detect"
          ]
        }
      ],
      "live": {
        "height": 720,
        "quality": 8
      },
      "motion": {
        "contour_area": 30,
        "delta_alpha": 0.2,
        "frame_alpha": 0.2,
        "frame_height": 50,
        "improve_contrast": false,
        "mask": "",
        "mqtt_off_delay": 30,
        "threshold": 25
      },
      "mqtt": {
        "bounding_box": true,
        "crop": true,
        "enabled": true,
        "height": 270,
        "quality": 70,
        "required_zones": [],
        "timestamp": true
      },
      "name": "alley_1",
      "objects": {
        "filters": {
          "person": {
            "mask": null,
            "max_area": 24000000,
            "max_ratio": 24000000,
            "min_area": 0,
            "min_ratio": 0,
            "min_score": 0.5,
            "threshold": 0.7
          }
        },
        "mask": "",
        "track": [
          "person"
        ]
      },
      "record": {
        "enabled": true,
        "events": {
          "objects": null,
          "post_capture": 5,
          "pre_capture": 5,
          "required_zones": [],
          "retain": {
            "default": 10,
            "mode": "motion",
            "objects": {}
          }
        },
        "expire_interval": 60,
        "retain": {
          "days": 0,
          "mode": "all"
        },
        "retain_days": null
      },
      "rtmp": {
        "enabled": true
      },
      "snapshots": {
        "bounding_box": true,
        "clean_copy": true,
        "crop": false,
        "enabled": true,
        "height": null,
        "quality": 70,
        "required_zones": [],
        "retain": {
          "default": 10,
          "mode": "motion",
          "objects": {}
        },
        "timestamp": false
      },
      "timestamp_style": {
        "color": {
          "blue": 255,
          "green": 255,
          "red": 255
        },
        "effect": null,
        "format": "%m/%d/%Y %H:%M:%S",
        "position": "tl",
        "thickness": 2
      },
      "ui": {
        "dashboard": true,
        "order": 0
      },
      "zones": {}
    }
  },
  "database": {
    "path": "/media/frigate/frigate.db"
  },
  "detect": {
    "enabled": true,
    "fps": 5,
    "height": 720,
    "max_disappeared": null,
    "stationary": {
      "interval": 0,
      "max_frames": {
        "default": null,
        "objects": {}
      },
      "threshold": null
    },
    "width": 1280
  },
  "detectors": {
    "cpu1": {
      "device": "usb",
      "num_threads": 2,
      "type": "cpu"
    }
  },
  "environment_vars": {},
  "ffmpeg": {
    "global_args": [
      "-hide_banner",
      "-loglevel",
      "warning"
    ],
    "hwaccel_args": [],
    "input_args": [
      "-avoid_negative_ts",
      "make_zero",
      "-fflags",
      "+genpts+discardcorrupt",
      "-rtsp_transport",
      "tcp",
      "-timeout",
      "5000000",
      "-use_wallclock_as_timestamps",
      "1"
    ],
    "output_args": {
      "detect": [
        "-f",
        "rawvideo",
        "-pix_fmt",
        "yuv420p"
      ],
      "record": [
        "-f",
        "segment",
        "-segment_time",
        "10",
        "-segment_format",
        "mp4",
        "-reset_timestamps",
        "1",
        "-strftime",
        "1",
        "-c",
        "copy",
        "-an"
      ],
      "rtmp": [
        "-c",
        "copy",
        "-f",
        "flv"
      ]
    }
  },
  "live": {
    "height": 720,
    "quality": 8
  },
  "logger": {
    "default": "info",
    "logs": {}
  },
  "model": {
    "height": 320,
    "labelmap": {},
    "labelmap_path": null,
    "path": null,
    "width": 320
  },
  "motion": null,
  "mqtt": {
    "client_id": "frigate",
    "host": "core-mosquitto",
    "password": <PASSWORD>,
    "port": 1883,
    "stats_interval": 60,
    "tls_ca_certs": null,
    "tls_client_cert": null,
    "tls_client_key": null,
    "tls_insecure": null,
    "topic_prefix": "frigate",
    "user": "frigate"
  },
  "objects": {
    "filters": null,
    "mask": "",
    "track": [
      "person"
    ]
  },
  "plus": {
    "enabled": false
  },
  "record": {
    "enabled": true,
    "events": {
      "objects": null,
      "post_capture": 5,
      "pre_capture": 5,
      "required_zones": [],
      "retain": {
        "default": 10,
        "mode": "motion",
        "objects": {}
      }
    },
    "expire_interval": 60,
    "retain": {
      "days": 0,
      "mode": "all"
    },
    "retain_days": null
  },
  "rtmp": {
    "enabled": true
  },
  "snapshots": {
    "bounding_box": true,
    "clean_copy": true,
    "crop": false,
    "enabled": false,
    "height": null,
    "quality": 70,
    "required_zones": [],
    "retain": {
      "default": 10,
      "mode": "motion",
      "objects": {}
    },
    "timestamp": false
  },
  "timestamp_style": {
    "color": {
      "blue": 255,
      "green": 255,
      "red": 255
    },
    "effect": null,
    "format": "%m/%d/%Y %H:%M:%S",
    "position": "tl",
    "thickness": 2
  },
  "ui": {
    "use_experimental": false
  }
}
```

**Comment by NickM-27** [comment]
as far as the ui lag goes, not totally sure why that is happening, but you should provide a full copy of your logs as that will likely have clues

the `detect_fps` is a live fps so it only will be more than 0 if it is currently detecting an object.

**Reply by Drizzt321** [reply]
@NickM-27 so I realized I should set the sub-stream (640x480@25fps) for the detect stream, and the main stream (1920x1080@5fps) as the record/rtmp stream. So reconfigured it that way, and the inference speed is now 53.98 (better) and shows camera fps as 25.5, so about right. 

However, the live view seems to be 4:3, or least that's how it's displaying. I think it's actually showing the sub-stream (640x480) as the main view, rather than the 1080p stream. 

Seems to be in a better place, so I'll see how it runs tomorrow. Right now the logs (through HA Add-On Logs tab) isn't showing anything as an issue.

**Reply by NickM-27** [reply]
> However, the live view seems to be 4:3, or least that's how it's displaying. I think it's actually showing the sub-stream (640x480) as the main view, rather than the 1080p stream.

Yes that is expected, there will be a new option in 0.12 to make the high res stream the live view.

**Reply by Drizzt321** [reply]
Ah, ok. Well that makes sufficient sense.

How can I increase the length of time before/after motion/object detected? Don't see any documentation for that, although from the Debug Config I tried changing these settings from 10 to 15, although that seemed to actually decrease the pre/post capture time?
```
record:
  events:
    pre_capture: 15
    post_capture: 15
```

Hmm...I do see some errors in the logs now, after it's been up for a while. I suspect this is something/bug from the IP camera. 
```
[2022-12-06 21:43:05] ffmpeg.alley_1.detect          ERROR   : [swscaler @ 0x559b2081f280] [swscaler @ 0x559b20e0d040] deprecated pixel format used, make sure you did set range correctly
[2022-12-06 21:43:05] ffmpeg.alley_1.detect          ERROR   : More than 1000 frames duplicated
[2022-12-06 21:43:05] ffmpeg.alley_1.detect          ERROR   : [h264 @ 0x559b208c0a80] error while decoding MB 14 27, bytestream -8
[2022-12-06 21:43:05] ffmpeg.alley_1.detect          ERROR   : rtsp://admin:<PW>@10.30.255.244/stream1: corrupt decoded frame in stream 0
[2022-12-06 21:43:05] ffmpeg.alley_1.record_rtmp     ERROR   : [segment @ 0x563d2cbf13c0] Timestamps are unset in a packet for stream 0. This is deprecated and will stop working in the future. Fix your code to set the timestamps properly
[2022-12-06 21:43:05] ffmpeg.alley_1.record_rtmp     ERROR   : [flv @ 0x563d2cbef840] Timestamps are unset in a packet for stream 0. This is deprecated and will stop working in the future. Fix your code to set the timestamps properly
[2022-12-06 21:43:05] ffmpeg.alley_1.record_rtmp     ERROR   : [segment @ 0x563d2cbf13c0] Non-monotonous DTS in output stream 0:0; previous: 0, current: 0; changing to 1. This may result in incorrect timestamps in the output file.
[2022-12-06 21:43:05] ffmpeg.alley_1.record_rtmp     ERROR   : [NULL @ 0x563d2cbdb000] missing picture in access unit with size 52
[2022-12-06 21:43:05] ffmpeg.alley_1.record_rtmp     ERROR   : [flv @ 0x563d2cbef840] Failed to update header with correct duration.
[2022-12-06 21:43:05] ffmpeg.alley_1.record_rtmp     ERROR   : [flv @ 0x563d2cbef840] Failed to update header with correct filesize.
[2022-12-06 21:43:05] watchdog.alley_1               INFO    : Terminating the existing ffmpeg process...
[2022-12-06 21:43:05] watchdog.alley_1               INFO    : Waiting for ffmpeg to exit gracefully...
```

**Reply by NickM-27** [reply]
The post_capture and pre_capture only apply within the segment. By default segments are ~10 seconds so any great value will have no affect. 

You can customize what is kept with recording and event retention modes https://docs.frigate.video/configuration/record#what-do-the-different-retain-modes-mean

**Reply by Drizzt321** [reply]
I'm a bit confused now. From re-reading some things on that link, I think I understand a bit better.

So Frigate only does 10 second recordings when it records. When playing back, Frigate concatenates any relevant 10 second segments together.

When an Object/Event occurs, it will almost certainly "overlap" 2 segments at a minimum, thus effectively 10 seconds before and 10 seconds after the event. So Frigate retains/records the 10 seconds before detection, and 10 seconds after (at a minimum). 

The Object/Event could persist across even more segments, so I could get a larger number of segments preserved/stored, and then played back in appropriate order based on Event/Object that I'm trying to view. 

So what does the pre_capture/post_capture do? And is there any way to increase segment size, at least a little bit? Or have it store more segments before/after Object/Event? Or does that not even make sense at all?

**Reply by NickM-27** [reply]
> So Frigate only does 10 second recordings when it records. When playing back, Frigate concatenates any relevant 10 second segments together.

correct

> When an Object/Event occurs, it will almost certainly "overlap" 2 segments at a minimum, thus effectively 10 seconds before and 10 seconds after the event. So Frigate retains/records the 10 seconds before detection, and 10 seconds after (at a minimum).
> 
> The Object/Event could persist across even more segments, so I could get a larger number of segments preserved/stored, and then played back in appropriate order based on Event/Object that I'm trying to view.

There could definitely be an object that is visible for less than 10 seconds and that happened to be right at the beginning of the segment, but yes you're right it is unlikely.

> So what does the pre_capture/post_capture do? And is there any way to increase segment size, at least a little bit? Or have it store more segments before/after Object/Event? Or does that not even make sense at all?

pre_capture & post_capture extend the time of the event in the relative directions. So if the object was visible at 9:10:00 AM and your pre_capture is 10 seconds then frigate will say that the event started at 9:09:50 AM. BUT whether or not that extra segment is kept depends on your retention mode. If you have an event retention mode of `all` then that extra segment will be kept, however if you have an event retention mode of `motion` and that segment from 9:09:50 to 9:10:00 had no motion then it won't be kept despite the pre_capture.

put simply, `pre_capture` and `post_capture` extends the amount of time that the event retention mode is applied, but they will not override that. You can set longer segments by adjusting the recording output_args but we don't recommend making it shorter or making it too long as it can have other consequences.

**Reply by Drizzt321** [reply]
Ahhh, ok, so they do do closer to what I thought they would, not quite, but closer. Now I just need to double check my event retention settings. 

So here's my settings. If I'm reading it right, that means it'll keep all segments for the last 5 days where it thinks there was any sort of motion in it.

And then, when an Event occurs, keep up to 10 days of segments, with when there was a detected Object that was active (e.g. not fully stationary). 

So the garbage bin just sitting there wouldn't trigger anything unless something actually jostled/moved it. In that case, if there was a detected Object (car, person, whatever) it would trigger an Event and record for however many segments there was a detected Object. Otherwise, it'd still be stored for up to 5 days (because of motion), but then how do I actually view that? The UI only lets me view by Event or Object, not full stream at a given time interval, regardless of detection.

```
record:
  enabled: true
  retain:
    days: 5
    mode: motion
  events:
    pre_capture: 20
    post_capture: 20
    retain:
      default: 10
      mode: active_objects
```

So for example, if I were to change the basic retain to `mode: 5`,  how would I view any part of the recording from 3 days ago, even if there was not Event or Object?

**Reply by NickM-27** [reply]
> The UI only lets me view by Event or Object, not full stream at a given time interval, regardless of detection.

that is incorrect, if you go to the cameras view you can open `recordings` which opens the 24/7 recordings page for that camera

**Reply by Drizzt321** [reply]
Hm, OK. Let me try changing the general retain to `all`, see how that affects the view. Since with `motion`, obviously I won't get the entirety of segments. 

Thanks for all the help! So far this seems pretty awesome, I'm thinking this will fulfill my personal needs :)
